InteractiveUtils: Support unannotated values after `Vararg` type annotation

serenity4 opened the pull request:
Implements #58900, ignoring any arguments coming after a `::Vararg`/`::Vararg{...}`/`::T...` annotation, after having ensured that no explicit type annotations were present on the dropped arguments.

The logic to detect whether it is a call to `Vararg` is a bit fishy, as we assume `A.B.C.Vararg` is the same as `Core.Vararg`, and we ignore that hygiene may apply another scope, but I think it is a reasonable behavior/expectation that if someone annotates an argument with`Vararg` they mean `Core.Vararg`, not a custom `Vararg` type. I don't feel strongly about that though.

serenity4 committed:
InteractiveUtils: Support unannotated values after Vararg type
annotation

serenity4 committed:
Merge branch 'master' of github.com:JuliaLang/julia into support-extended-varargs

topolarity commented on stdlib/InteractiveUtils/src/macros.jl:
```diff
@@ -9,8 +9,34 @@ using Base: insert!, replace_ref_begin_end!,
+        is_vararg = isexpr(arg, :...) && isexpr(arg.args[1], :(::)) ||
+                    # technically, a user may define their own `Vararg` binding,
+                    # but handling this right would require deferring the check until runtime,
+                    # where we can't differentiate between a type coming from a value or a type annotation
```
Can you explain more about this ambiguity?

`typeof` should never yield a (naked) `Vararg{...}` afaik

topolarity replied:
Oh! To implement:
> if you have explicit type-ascriptions after the Vararg, that should arguably be an error

I see.

As someone who has created a `Vararg` type in a package before (it was a partial implementation of Julia's subtyping, so not too surprising it showed up), I would like to avoid making syntactic assumptions on what "Vararg" means

If it's necessary, I think I am slightly more in favor of ignoring any types after the Vararg (explicit or not), if it means restoring that

serenity4 replied:
I moved to check to runtime in ac19f4ee8e94951d2355a86aaadaffa50e3bace5, so we no longer have to rely on these fishy assumptions. Do you like it better now?

serenity4 committed:
Move Vararg handling to runtime

This is to make the handling more accurate,
without having to assume that `Vararg` symbols
refer to `Core.Vararg`

topolarity commented on stdlib/InteractiveUtils/src/macros.jl:
```diff
@@ -9,8 +9,26 @@ using Base: insert!, replace_ref_begin_end!,
+    throw(ArgumentError("More than one `Core.Vararg` type present in argument tuple ($(join(types[varargs], ", "))); if provided, it must be unique"))
+end
+
+function is_vararg_symbol(@nospecialize ex)
+    ex = ignore_qualifiers(ex)
+    name = isexpr(ex, :curly) ? ex.args[1] : ex
+    return in(name, (:Vararg, Vararg))
+end
+ignore_qualifiers(@nospecialize ex) = isexpr(ex, :(.), 2) ? ex.args[2] : ex
```
Looks like these are unused now?

serenity4 replied:
Oh, right!

topolarity reviewed (approved):
Looks great, thanks!

FWIW if you want to detect some errors still, I think it'd be correct to test:
1. `Tᵢ <: Base.unwrapva(VT)` for all following types
2. if `Vararg{T,N}`, you could check that exactly N-1 types appear afterward

which would reject some of the "obviously inconsistent" examples in the tests

serenity4 committed:
Remove unused code

serenity4 committed:
Improve accuracy of checks

serenity4 committed:
Remove extra parentheses

topolarity commented:
The counting seems a bit off to my eyes:
```julia
 julia> foo(x::Vararg{Int,3}) = +(x...)
 julia> foo(1, 2, 3)
 6
 
 julia> @code_typed foo(::Vararg{Int,3}) # I'd expect this to be interpreted like a splat
ERROR: ArgumentError: Expected exactly 3 types after `Vararg{Int64, 3}`, found 0 instead 

 julia> @code_typed foo(1::Vararg{Int,3}, 2, 3)
ERROR: ArgumentError: Expected exactly 3 types after `Vararg{Int64, 3}`, found 0 instead

 julia> @code_typed foo(1::Vararg{Int,3}, 2, 3, 4)
 # succeeds?
```

also, bit of a strange corner-case but:
```julia
julia> @code_typed +(::T, ::Vararg{T}, ::T) where T
ERROR: TypeError: in <:, expected Type, got a value of type TypeVar
```

The subtyping check for this would require prenex polymorphism (i.e. checking whether this is true ∀T), which Julia doesn't have implemented, so most likely the check should just be disabled in this case.

serenity4 committed:
Fixes, don't check type for UnionAll/TypeVar types

topolarity commented on stdlib/InteractiveUtils/src/macros.jl:
```diff
@@ -9,8 +9,33 @@ using Base: insert!, replace_ref_begin_end!,
+    return Tuple{@view(types[1:vararg])...}
+end
+
+skip_type_check(@nospecialize(T)) = isa(T, TypeVar) || isa(T, UnionAll) || Core.has_free_typevars(T)
```
Is the UnionAll check necessary? I'd expect the last bit to be enough on its own

serenity4 committed:
Type check UnionAll too

topolarity reviewed (approved):
LGTM! Let's merge it.

serenity4 committed:
Update stdlib/InteractiveUtils/src/macros.jl

Co-authored-by: Cody Tapscott <84105208+topolarity@users.noreply.github.com>
